src/Language/Haskell/TH/Datatype.hs

```diff
@@ -833,6 +836,15 @@ stripSigT :: Type -> Type
 stripSigT (SigT t _) = t
 stripSigT t          = t
 
+#if MIN_VERSION_template_haskell(2,16,0)
+-- | Get the 'Kind' in the tail position of a 'Kind'.
+-- Look inside any quantifier, and for the target of an arrow.
+getRetKind :: Kind -> Kind
+getRetKind (ForallT vars _ r) = getRetKind r
+getRetKind (ForallVisT vars r) = getRetKind r
+getRetKind (AppT (AppT ArrowT _) r) = getRetKind r
+getRetKind r = r
```

Comment:
I don't think that this will work if the return kind mentions a kind synonym, e.g.,

```hs
type TT = Type -> Type

data D :: TT where ...
```

We should be able to fix this by calling [`resolveKindSynonyms`](https://hackage.haskell.org/package/th-abstraction-0.6.0.0/docs/Language-Haskell-TH-Datatype.html#v:resolveKindSynonyms) beforehand. We already call `resolveKindSynonyms` [here](https://github.com/glguy/th-abstraction/blob/8130497fc938cbb96dac40047617aaa85911f9b5/src/Language/Haskell/TH/Datatype.hs#L805) in `mkExtraKindBinders`, so perhaps we can plumb the return kind from that function (which is already free of kind synonyms) to `normalizeDec'`.

This would require a bit of refactoring to accomplish, but as I mention in my earlier comment, I think we should perform a refactoring anyway.

Reply:
That sounds good! I'll refactor things to reuse the return kind calculated by `mkExtraKindBinders`

Reply:
This function is no longer used, so let's remove it.